provider/datagen/src/transform/cldr/list/mod.rs

```diff
@@ -3,164 +3,147 @@
 // (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
 
 use crate::transform::cldr::cldr_serde;
-use crate::transform::icuexport::uprops::EnumeratedPropertyCodePointTrieProvider;
-use crate::SourceData;
 use icu_list::provider::*;
 use icu_locid::subtags_language as language;
 use icu_provider::datagen::IterableDataProvider;
 use icu_provider::prelude::*;
 use lazy_static::lazy_static;
 
-/// A data provider reading from CLDR JSON list rule files.
-#[derive(Debug)]
-pub struct ListProvider {
-    source: SourceData,
-}
-
-impl From<&SourceData> for ListProvider {
-    fn from(source: &SourceData) -> Self {
-        ListProvider {
-            source: source.clone(),
-        }
-    }
-}
-
-impl<M: KeyedDataMarker<Yokeable = ListFormatterPatternsV1<'static>>> DataProvider<M>
-    for ListProvider
-{
-    fn load(&self, req: DataRequest) -> Result<DataResponse<M>, DataError> {
-        let langid = req.locale.get_langid();
-
-        let resource: &cldr_serde::list_patterns::Resource = self
-            .source
-            .cldr()?
-            .misc()
-            .read_and_parse(&langid, "listPatterns.json")?;
-
-        let data = &resource
-            .main
-            .0
-            .get(&langid)
-            .expect("CLDR file contains the expected language")
-            .list_patterns;
-
-        let (wide, short, narrow) = if M::KEY == AndListV1Marker::KEY {
-            (&data.standard, &data.standard_short, &data.standard_narrow)
-        } else if M::KEY == OrListV1Marker::KEY {
-            (&data.or, &data.or_short, &data.or_narrow)
-        } else if M::KEY == UnitListV1Marker::KEY {
-            (&data.unit, &data.unit_short, &data.unit_narrow)
-        } else {
-            return Err(DataError::custom("Unknown key for ListFormatterPatternsV1"));
-        };
-
-        let mut patterns = ListFormatterPatternsV1::try_new([
-            &wide.start,
-            &wide.middle,
-            &wide.end,
-            &wide.pair,
-            &short.start,
-            &short.middle,
-            &short.end,
-            &short.pair,
-            &narrow.start,
-            &narrow.middle,
-            &narrow.end,
-            &narrow.pair,
-        ])?;
-
-        if langid.language == language!("es") {
-            if M::KEY == AndListV1Marker::KEY || M::KEY == UnitListV1Marker::KEY {
-                lazy_static! {
-                    // Starts with i or (hi but not hia/hie)
-                    static ref I_SOUND: StringMatcher<'static> = StringMatcher::new("i|hi([^ae]|$)").expect("Valid regex");
+macro_rules! implement {
```

Comment:
Nit: I really prefer keeping code out of macros, because code analysis tools like clippy, rust-analyzer, and rustfmt do not work well inside macros. Can you try to move as much of this code as possible into a helper function and only keep a minimal amount inside the macro?